# L2 Escape Model

Consequent Reversibility was not included in the final escape models because, in this subgroup structure, it is partially encoded by Goal_Type and produced unstable estimates when entered together with Goal_Type.

Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Overall Subgroup:
1. Create escape_L2
2. Counts and Proportions
3. Logistic Models
4. Model Comparison

Read data

In [2]:
subgroup_theoretical = pd.read_csv("../../data/processed/subgroup_theoretical.csv")

1. Create escape_2

In [3]:
subgroup_theoretical["escape_L2"] = (
    subgroup_theoretical["Response_Full"] != "L2_other"
).astype(int)

2. Counts and Proportions

In [4]:
escape_L2_counts = pd.crosstab(
    subgroup_theoretical["escape_L2"],
    subgroup_theoretical["Response_Full"])

escape_L2_counts

Response_Full,L1_transfer,L2_other,correct
escape_L2,,,
0,0,38,0
1,96,0,61


In [5]:
escape_L2_props = subgroup_theoretical["escape_L2"].value_counts(normalize = True)
escape_L2_props

escape_L2
1    0.808081
0    0.191919
Name: proportion, dtype: float64

Counts and Proportions by Condition

In [6]:
counts_condition = pd.crosstab(
    [subgroup_theoretical["Goal_Type"], subgroup_theoretical["Agent"], subgroup_theoretical["Consequent_Reversibility"]],
    subgroup_theoretical["escape_L2"],
    margins = True)

counts_condition

escape_L2                                          0    1  All
Goal_Type         Agent Consequent_Reversibility              
goal_frequent     0     Yes                       14   28   42
                  1     Yes                        2   22   24
goal_non_frequent 0     No                         9   33   42
                  1     No                         1   23   24
no_goal           0     Yes                        8   34   42
                  1     Yes                        4   20   24
All                                               38  160  198

In [7]:
props_condition = pd.crosstab(
    [subgroup_theoretical["Goal_Type"], subgroup_theoretical["Agent"], subgroup_theoretical["Consequent_Reversibility"]],
    subgroup_theoretical["escape_L2"],
    normalize= "index")

props_condition

escape_L2                                                0         1
Goal_Type         Agent Consequent_Reversibility                    
goal_frequent     0     Yes                       0.333333  0.666667
                  1     Yes                       0.083333  0.916667
goal_non_frequent 0     No                        0.214286  0.785714
                  1     No                        0.041667  0.958333
no_goal           0     Yes                       0.190476  0.809524
                  1     Yes                       0.166667  0.833333

Agency appears to help participants escape L2. 

The condition with the strongest L1 transfer (goal_non_frequent, agent=1) also exhibits the highest L2 escape rate (95.8%). This suggests that failure does not occur at the L2 suppression stage, but at the subsequent resolution stage.

In addition, if there is a goal representation available, agency seems to help participants commit by leaving the L2 space; however, if there is no goal representation, agency seems to have much less to work with. 

3. Logistic Models

Import Libraries

In [8]:
import statsmodels.formula.api as smf
from scipy.stats import chi2

Binary Outcome

In [9]:
subgroup_theoretical["escape_L2"] = (
    subgroup_theoretical["Response_Full"] != "L2_other"
).astype(int)

In [10]:
subgroup_theoretical["escape_L2"].value_counts()

escape_L2
1    160
0     38
Name: count, dtype: int64

**Model E1**
escape_L2 ~ Agent

In [35]:
E1 = smf.logit(
    "escape_L2 ~ Agent",
    data=subgroup_theoretical
    ).fit()

print(E1.summary())

Optimization terminated successfully.
         Current function value: 0.471025
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      196
Method:                           MLE   Df Model:                            1
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.03675
Time:                        17:31:05   Log-Likelihood:                -93.263
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                  0.007642
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.1199      0.207      5.414      0.000       0.714       1.525
Agent          1.1086      0.

**Model E2**
escape_L2 ~ Agent * Goal_Type

In [36]:
E2 = smf.logit(
    "escape_L2 ~ Agent * Goal_Type",
    data= subgroup_theoretical
    ).fit()

print(E2.summary())

Optimization terminated successfully.
         Current function value: 0.458893
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      192
Method:                           MLE   Df Model:                            5
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.06156
Time:                        17:32:30   Log-Likelihood:                -90.861
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                   0.03590
                                           coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                                0.6931      0.3

additive

In [65]:
E3 = smf.logit(
    "escape_L2 ~ Agent + Goal_Type",
    data= subgroup_theoretical
    ).fit()

print(E3.summary())

Optimization terminated successfully.
         Current function value: 0.466314
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      194
Method:                           MLE   Df Model:                            3
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.04638
Time:                        18:08:56   Log-Likelihood:                -92.330
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                   0.02954
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          0.8081      0.311      2.599     

In [66]:
E4 = smf.logit(
    "escape_L2 ~ Agent + Consequent_Reversibility",
    data= subgroup_theoretical
    ).fit()

print(E4.summary())

Optimization terminated successfully.
         Current function value: 0.468220
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      195
Method:                           MLE   Df Model:                            2
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.04248
Time:                        18:19:24   Log-Likelihood:                -92.708
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                   0.01636
                                      coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                           1.4127      0.359      3.932  

**Model E3**
escape_L2 ~ Consequent_Reversibility + Goal_Type + Agent

In [17]:
from scipy.stats import chi2

def lr_test(model_small, model_large):

    LR = 2 * (model_large.llf - model_small.llf)

    df = model_large.df_model - model_small.df_model

    p = chi2.sf(LR, df)

    return pd.Series({
        "LR": LR,
        "df": df,
        "p_value": p
    })

In [39]:
lr_test(E1, E2)

LR         4.804579
df         4.000000
p_value    0.307943
dtype: float64

**Resolve_correct = "correct"**

In [69]:
escapees = subgroup_theoretical[subgroup_theoretical["escape_L2"] == 1].copy()

In [70]:
escapees["resolved_correct"] = (escapees["Response_Full"] == "correct").astype(int)

In [71]:
R1 = smf.logit(
    "resolved_correct ~ Agent",
    data= escapees
    ).fit()

print(R1.summary())

Optimization terminated successfully.
         Current function value: 0.664164
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:       resolved_correct   No. Observations:                  160
Model:                          Logit   Df Residuals:                      158
Method:                           MLE   Df Model:                            1
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:               0.0007657
Time:                        18:49:29   Log-Likelihood:                -106.27
converged:                       True   LL-Null:                       -106.35
Covariance Type:            nonrobust   LLR p-value:                    0.6865
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.5390      0.213     -2.534      0.011      -0.956      -0.122
Agent          0.1335      0.

In [73]:
R2 = smf.logit(
    "resolved_correct ~ Goal_Type",
    data= escapees
    ).fit()

print(R2.summary())

Optimization terminated successfully.
         Current function value: 0.655378
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:       resolved_correct   No. Observations:                  160
Model:                          Logit   Df Residuals:                      157
Method:                           MLE   Df Model:                            2
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.01398
Time:                        18:50:09   Log-Likelihood:                -104.86
converged:                       True   LL-Null:                       -106.35
Covariance Type:            nonrobust   LLR p-value:                    0.2260
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                         -0.0800      0.283     -0.283     

In [74]:
R3 = smf.logit(
    "resolved_correct ~ Consequent_Reversibility",
    data= escapees
    ).fit()

print(R3.summary())

Optimization terminated successfully.
         Current function value: 0.662647
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:       resolved_correct   No. Observations:                  160
Model:                          Logit   Df Residuals:                      158
Method:                           MLE   Df Model:                            1
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                0.003047
Time:                        18:51:00   Log-Likelihood:                -106.02
converged:                       True   LL-Null:                       -106.35
Covariance Type:            nonrobust   LLR p-value:                    0.4208
                                      coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          -0.6665      0.282     -2.361  

In [75]:
R4 = smf.logit(
    "resolved_correct ~ Goal_Type + Agent + Consequent_Reversibility",
    data= escapees
    ).fit()

print(R4.summary())

         Current function value: 0.655049
         Iterations: 35
                           Logit Regression Results                           
Dep. Variable:       resolved_correct   No. Observations:                  160
Model:                          Logit   Df Residuals:                      156
Method:                           MLE   Df Model:                            3
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.01448
Time:                        18:51:42   Log-Likelihood:                -104.81
converged:                      False   LL-Null:                       -106.35
Covariance Type:            nonrobust   LLR p-value:                    0.3795
                                      coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          -0.2798    1.5e+07  -1.86e-08      1.000   -2.94e+07    2.94e+07
Go

/opt/anaconda3/envs/counterfactual-env/lib/python3.12/site-packages/statsmodels/base/model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


In [76]:
R5 = smf.logit(
    "resolved_correct ~ Goal_Type * Agent",
    data= escapees
    ).fit()

print(R5.summary())

Optimization terminated successfully.
         Current function value: 0.648393
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:       resolved_correct   No. Observations:                  160
Model:                          Logit   Df Residuals:                      154
Method:                           MLE   Df Model:                            5
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.02449
Time:                        18:52:50   Log-Likelihood:                -103.74
converged:                       True   LL-Null:                       -106.35
Covariance Type:            nonrobust   LLR p-value:                    0.3909
                                           coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                            -5.629e-17      0.3

Unresolve_L1

In [52]:
U1 = smf.logit(
    "unresolve_L1 ~ Agent",
    data= df_conflict
    ).fit()

print(U1.summary())

Optimization terminated successfully.
         Current function value: 0.667893
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:           unresolve_L1   No. Observations:                  157
Model:                          Logit   Df Residuals:                      155
Method:                           MLE   Df Model:                            1
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:               0.0002923
Time:                        17:49:44   Log-Likelihood:                -104.86
converged:                       True   LL-Null:                       -104.89
Covariance Type:            nonrobust   LLR p-value:                    0.8044
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.4877      0.215      2.271      0.023       0.067       0.909
Agent         -0.0822      0.

In [53]:
U2 = smf.logit(
    "unresolve_L1 ~ Agent * Goal_Type",
    data= df_conflict
    ).fit()

print(U2.summary())

Optimization terminated successfully.
         Current function value: 0.652192
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:           unresolve_L1   No. Observations:                  157
Model:                          Logit   Df Residuals:                      151
Method:                           MLE   Df Model:                            5
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.02379
Time:                        17:49:57   Log-Likelihood:                -102.39
converged:                       True   LL-Null:                       -104.89
Covariance Type:            nonrobust   LLR p-value:                    0.4169
                                           coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                               -0.0741      0.3

## Actors Subgroup:
1. Create escape_L2
2. Counts and Proportions
3. Logistic Models
4. Model Comparison

Read data

In [19]:
subgroup_actors = subgroup_theoretical[subgroup_theoretical["Focus"] == "I"].copy()

1. Create escape_L2

In [20]:
subgroup_actors["escape_L2"] = (
    subgroup_actors["Response_Full"] != "L2_other"
).astype(int)

2. Counts and Proportions

In [21]:
escape_L2_counts = pd.crosstab(
    subgroup_actors["escape_L2"],
    subgroup_actors["Response_Full"])

escape_L2_counts

Response_Full,L1_transfer,L2_other,correct
escape_L2,,,
0,0,24,0
1,53,0,40


In [22]:
escape_L2_props = subgroup_actors["escape_L2"].value_counts(normalize = True)
escape_L2_props

escape_L2
1    0.8
0    0.2
Name: proportion, dtype: float64

Counts and Proportions by Condition

In [23]:
counts_condition = pd.crosstab(
    [subgroup_actors["Goal_Type"], subgroup_actors["Agent"], subgroup_actors["Consequent_Reversibility"]],
    subgroup_actors["escape_L2"],
    margins = True)

counts_condition

escape_L2                                          0   1  All
Goal_Type         Agent Consequent_Reversibility             
goal_frequent     0     Yes                        8  16   24
                  1     Yes                        2  14   16
goal_non_frequent 0     No                         8  16   24
                  1     No                         1  15   16
no_goal           0     Yes                        3  21   24
                  1     Yes                        2  14   16
All                                               24  96  120

In [24]:
props_condition = pd.crosstab(
    [subgroup_actors["Goal_Type"], subgroup_actors["Agent"], subgroup_actors["Consequent_Reversibility"]],
    subgroup_actors["escape_L2"],
    normalize= "index")

props_condition

escape_L2                                                0         1
Goal_Type         Agent Consequent_Reversibility                    
goal_frequent     0     Yes                       0.333333  0.666667
                  1     Yes                       0.125000  0.875000
goal_non_frequent 0     No                        0.333333  0.666667
                  1     No                        0.062500  0.937500
no_goal           0     Yes                       0.125000  0.875000
                  1     Yes                       0.125000  0.875000

3. Logistic Models

Binary Outcome

In [25]:
subgroup_actors["escape_L2"] = (
    subgroup_actors["Response_Full"] != "L2_other"
).astype(int)

In [26]:
subgroup_actors["escape_L2"].value_counts()

escape_L2
1    96
0    24
Name: count, dtype: int64

4. Model comparisons

**E1 Model**: 
escape_L2 ~ Consequent_Reversibility

In [55]:
E1_A = smf.logit(
    "escape_L2 ~ Agent",
    data=subgroup_actors
    ).fit()

print(E1_A.summary())

Optimization terminated successfully.
         Current function value: 0.479908
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      118
Method:                           MLE   Df Model:                            1
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.04096
Time:                        17:51:01   Log-Likelihood:                -57.589
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                   0.02657
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.0259      0.267      3.836      0.000       0.502       1.550
Agent          1.1259      0.

For actors, consequent reversibility did not significantly predict L2 escape (p = .629), suggesting that reversibility alone is insufficient to trigger abandonment of L2 interpretations. This supports the view that reversibility functions as a source of latent cognitive load whose effects become observable only when additional representational demands are introduced.

In [56]:
E2_A = smf.logit(
    "escape_L2 ~ Agent * Goal_Type",
    data=subgroup_actors
    ).fit()

print(E2_A.summary())

Optimization terminated successfully.
         Current function value: 0.461604
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      114
Method:                           MLE   Df Model:                            5
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.07753
Time:                        17:51:42   Log-Likelihood:                -55.392
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                   0.09726
                                           coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                                0.6931      0.4

In [67]:
E3_A = smf.logit(
    "escape_L2 ~ Agent + Goal_Type",
    data=subgroup_actors
    ).fit()

print(E3_A.summary())

Optimization terminated successfully.
         Current function value: 0.469898
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      116
Method:                           MLE   Df Model:                            3
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.06096
Time:                        18:26:19   Log-Likelihood:                -56.388
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                   0.06234
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          0.7203      0.400      1.799     

In [68]:
E4_A = smf.logit(
    "escape_L2 ~ Agent + Consequent_Reversibility",
    data=subgroup_actors
    ).fit()

print(E4_A.summary())

Optimization terminated successfully.
         Current function value: 0.478907
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      117
Method:                           MLE   Df Model:                            2
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.04296
Time:                        18:27:19   Log-Likelihood:                -57.469
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                   0.07581
                                      coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                           0.8695      0.410      2.118  

Resolve_L1

In [59]:
df_conflict_a = subgroup_actors[subgroup_actors['Response_Full'].isin(['correct', 'L1_transfer'])].copy()

In [60]:
df_conflict_a['resolve_L1'] = (df_conflict_a['Response_Full'] == 'correct').astype(int)
df_conflict_a['unresolve_L1']= (df_conflict_a['Response_Full'] == 'L1_transfer').astype(int)

In [61]:
R1_A = smf.logit(
    "resolve_L1 ~ Agent",
    data= df_conflict_a
    ).fit()

print(R1_A.summary())

Optimization terminated successfully.
         Current function value: 0.677419
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:             resolve_L1   No. Observations:                   93
Model:                          Logit   Df Residuals:                       91
Method:                           MLE   Df Model:                            1
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                0.008673
Time:                        17:56:28   Log-Likelihood:                -63.000
converged:                       True   LL-Null:                       -63.551
Covariance Type:            nonrobust   LLR p-value:                    0.2938
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.0800      0.283     -0.283      0.777      -0.635       0.475
Agent         -0.4432      0.

In [62]:
R2_A = smf.logit(
    "resolve_L1 ~ Agent * Goal_Type",
    data= df_conflict_a
    ).fit()

print(R2_A.summary())

Optimization terminated successfully.
         Current function value: 0.649860
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:             resolve_L1   No. Observations:                   93
Model:                          Logit   Df Residuals:                       87
Method:                           MLE   Df Model:                            5
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.04900
Time:                        17:56:58   Log-Likelihood:                -60.437
converged:                       True   LL-Null:                       -63.551
Covariance Type:            nonrobust   LLR p-value:                    0.2846
                                           coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                                0.4055      0.5

unresolved_L1 = L1_transfer

In [63]:
U1_A = smf.logit(
    "unresolve_L1 ~ Agent",
    data= df_conflict_a
    ).fit()

print(U1_A.summary())

Optimization terminated successfully.
         Current function value: 0.677419
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:           unresolve_L1   No. Observations:                   93
Model:                          Logit   Df Residuals:                       91
Method:                           MLE   Df Model:                            1
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                0.008673
Time:                        17:58:48   Log-Likelihood:                -63.000
converged:                       True   LL-Null:                       -63.551
Covariance Type:            nonrobust   LLR p-value:                    0.2938
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0800      0.283      0.283      0.777      -0.475       0.635
Agent          0.4432      0.

In [64]:
U2_A = smf.logit(
    "unresolve_L1 ~ Agent * Goal_Type",
    data= df_conflict_a
    ).fit()

print(U2_A.summary())

Optimization terminated successfully.
         Current function value: 0.649860
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:           unresolve_L1   No. Observations:                   93
Model:                          Logit   Df Residuals:                       87
Method:                           MLE   Df Model:                            5
Date:                Tue, 16 Jun 2026   Pseudo R-squ.:                 0.04900
Time:                        17:59:18   Log-Likelihood:                -60.437
converged:                       True   LL-Null:                       -63.551
Covariance Type:            nonrobust   LLR p-value:                    0.2846
                                           coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                               -0.4055      0.5

# Short Comparison:
1. Does the same predictor structure hold?

- For the Escape_L2 stage: Logistic regression analyses of L2 escape showed a consistent main effect of Agency across both the theoretical and actor subgroups. Neither Goal Type nor Consequent Reversibility improved model fit or attenuated the Agency coefficient, suggesting that escape from L2 is primarily associated with the presence of agency representations.

2. Does effect size/ model fit improve?